src/RadixDlt.NetworkGateway.PostgresIntegration/Queries/NonFungibleVaultContentsQuery.cs

```diff
@@ -0,0 +1,247 @@
+        LIMIT var.scan_limit
+    ) ed
+    INNER JOIN LATERAL (
+        SELECT
+            *,
+            CASE WHEN var.ignore_deleted THEN NOT is_deleted ELSE TRUE END AS should_be_returned
+        FROM non_fungible_vault_entry_history
+        WHERE non_fungible_vault_entry_definition_id = ed.id AND ed.from_state_version <= var.at_ledger_state
+        ORDER BY from_state_version DESC
+        LIMIT 1
+    ) eh ON TRUE
+    WHERE ed.is_last_candidate OR eh.should_be_returned
+    ORDER BY
+        CASE WHEN var.descending_order THEN ed.from_state_version END DESC,
+        CASE WHEN var.descending_order THEN ed.id END DESC,
+        CASE WHEN NOT var.descending_order THEN ed.from_state_version END,
+        CASE WHEN NOT var.descending_order THEN ed.id END
+    LIMIT var.page_limit
+) entries ON TRUE
+INNER JOIN non_fungible_id_definition nf_def ON nf_def.id = entries.non_fungible_id_definition_id
+LEFT JOIN LATERAL (
+    SELECT *
+    FROM non_fungible_id_data_history
+    WHERE non_fungible_id_definition_id = nf_def.id AND from_state_version <= var.at_ledger_state
+    ORDER BY from_state_version DESC
+    LIMIT 1
+) nf_data ON var.include_data
+LEFT JOIN LATERAL (
+    SELECT *
+    FROM vault_balance_history
+    WHERE vault_entity_id = var.vault_entity_id AND from_state_version <= var.at_ledger_state
+    ORDER BY from_state_version DESC
+    LIMIT 1
+) balance ON TRUE
+ORDER BY
+    CASE WHEN var.descending_order THEN entries.definition_from_state_version END DESC,
+    CASE WHEN var.descending_order THEN entries.non_fungible_id_definition_id END DESC,
+    CASE WHEN NOT var.descending_order THEN entries.definition_from_state_version END,
+    CASE WHEN NOT var.descending_order THEN entries.non_fungible_id_definition_id END;",
+            new
+            {
+                vaultEntityIds = vaultEntityIds.ToList(),
+                scanLimit = 5000 + 1, // TODO configurable / use const
+                pageLimit = configuration.NonFungibleIdsPerVault + 1,
+                descendingOrder = configuration.DescendingOrder,
+                vaultCursorStateVersion = configuration.Cursor?.StateVersion ?? (configuration.DescendingOrder ? long.MaxValue : long.MinValue),
+                vaultCursorId = configuration.Cursor?.Id ?? (configuration.DescendingOrder ? long.MaxValue : long.MinValue),
+                atLedgerState = configuration.AtLedgerState,
+            },
+            token);
+
+        var result = new Dictionary<long, ResultVault>();
+
+        await dapperWrapper.QueryAsync<ResultVault, ResultNonFungibleId, ResultVault>(
+            dbContext.Database.GetDbConnection(),
+            cd,
+            (vaultRow, entryRow) =>
+            {
+                var vault = result.GetOrAdd(vaultRow.VaultEntityId, _ => vaultRow);
+
+                if (entryRow.DefinitionIsLastCandidate || vault.NonFungibleIds.Count >= configuration.NonFungibleIdsPerVault)
```

Comment:
> Why >= rather than == out of interest? Is there a time when we expect to return more than the limit?

There should ever only be _at most one_ cursor available. Should there ever be another cursor it would indicate something being incorrect with the query. My thinking was that `>=` combined with `public StateVersionIdCursor? NonFungibleIdsNextCursor { get; set; } // TODO set should throw if already non-null` should result in pretty exception and a crash. IMO that's much better than the alternative, i.e. to silently ignore the cursor overwrite and/or add elements to the resultant collection past the max page size.